src/filters/PoseEstimator.cpp

```diff
@@ -1,49 +1,124 @@
 #include "PoseEstimator.h"
 
-typedef Eigen::Matrix3d matrix;
-typedef Eigen::Vector3d vector;
-
-PoseEstimator::PoseEstimator(const Eigen::Vector3d &stateStdDevs,
-							 const Eigen::Vector3d &measurementStdDevs, double dt)
-	: kf(KalmanFilter<3, 3>::createDiscrete(matrix::Identity(), matrix::Identity(),
-										matrix::Identity(), stateStdDevs, measurementStdDevs,
-										dt)),
-	  dt(dt)
+// Note: The state vector is defined as [x, y, theta].
+// The input vector is defined as [lVel, rVel].
+
+namespace
+{
+statevec_t stateFunc(double wheelBase, double dt, const statevec_t &x, const Eigen::Vector2d &u,
+					 const Eigen::Vector2d &w)
 {
+	// Instead of using euler integration, we can represent the pose update as a twist
+	// and then apply that twist to the current pose. Derived from differential drive kinematics.
+	// This will be more accurate than euler integration at slower update rates.
+	Eigen::Vector2d input = u + w; // noise is applied to the input vector
+	double theta = x(2);
+
+	Eigen::Matrix<double, 3, 2> trf;
+	trf << 0.5, 0.5, 0, 0, -1 / wheelBase, 1 / wheelBase;
+
+	// This represents [deltaX, 0, deltaTheta]
+	// however, this is "pre-transform". We will now transform it to give a more accurate
+	// deltaX and deltaY.
+	Eigen::Vector3d localUpdate = trf * input * dt;
+
+	double cosTheta = cos(theta);
+	double sinTheta = sin(theta);
+	// This matrix transforms from local space to map space.
+	Eigen::Matrix3d A;
+	A << cosTheta, -sinTheta, 0, sinTheta, cosTheta, 0, 0, 0, 1;
+
+	double dTheta = localUpdate(2);
+	Eigen::Matrix3d B;
+	// If dTheta is close to 0 use the taylor series approximation
+	if (abs(dTheta) <= 1e-9)
+	{
+		B << 1 - dTheta * dTheta / 6.0, -dTheta / 2.0, 0, dTheta / 2.0,
+			1 - dTheta * dTheta / 6.0, 0, 0, 0, 1;
+	}
+	else
+	{
+		double sinDTheta = sin(dTheta);
+		double cosDTheta = cos(dTheta);
+		B << sinDTheta, cosDTheta - 1, 0, 1 - cosDTheta, sinDTheta, 0, 0, 0, dTheta;
+		B /= dTheta;
+	}
+
+	return x + A * B * localUpdate;
```

Comment:
I haven't gone through this math in detail, so I guess I'll just run the code and see if it behaves reasonably. I'm not too familiar with analytic integration of differential drive kinematics. Sounds interesting though. . . .

Math like this might be a good case for adding some automated tests. E.g. if we start at pose (0,0,0) and apply velocity 1m/s to both wheels for half a second, we should end at pose (0.5,0,0). And some more complicated tests that involve turning, going backward, etc.